### ISSUE TITLE
Invalid error message for E0277

### ISSUE THREAD

**Issue by l4l** [rust-lang/rust#57800]
When trying to use const-fn at trait signature, e.g via function return array of fixed size:

```rust
trait T {
    fn f() -> [u8; std::mem::size_of::<Self>()];
}
```

Compiler seem to produce in some terms incorrect error:
```
error[E0277]: the size for values of type `Self` cannot be known at compilation time
 --> src/main.rs:3:20
  |
3 |     fn f() -> [u8; std::mem::size_of::<Self>()]
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `std::marker::Sized` is not implemented for `Self`
  = note: to learn more, visit <https://doc.rust-lang.org/book/second-edition/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
  = help: consider adding a `where Self: std::marker::Sized` bound
  = note: required by `std::mem::size_of`

error: aborting due to previous error
```

And in fact, adding trait bound [won't help](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3a8730055aae42a36349579e278bdb1e) and will lead to the same error. Probably that should lead to: "Const generics are not supported", or might something more precise.

**Comment by kadiwa4** [comment]
(nightly 2023-01-23) It now gives you a generic const parameters error:
```
error: generic parameters may not be used in const operations
 --> src/lib.rs:2:40
  |
2 |     fn f() -> [u8; std::mem::size_of::<Self>()];
  |                                        ^^^^ cannot perform const operation using `Self`
  |
  = note: type parameters may not be used in const expressions
  = help: use `#![feature(generic_const_exprs)]` to allow generic const expressions

error: could not compile `playground` due to previous error
```
After adding the `#![feature]`, it gives you a suggestion which works:
```
error[E0277]: the size for values of type `Self` cannot be known at compilation time
 --> src/lib.rs:5:40
  |
5 |     fn f() -> [u8; std::mem::size_of::<Self>()];
  |                                        ^^^^ doesn't have a size known at compile-time
  |
note: required by a bound in `std::mem::size_of`
 --> /rustc/c8e6a9e8b6251bbc8276cb78cabe1998deecbed7/library/core/src/mem/mod.rs:308:1
help: consider further restricting `Self`
  |
5 |     fn f() -> [u8; std::mem::size_of::<Self>()] where Self: Sized;
  |                                                 +++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to previous error
```
@rustbot label: +E-needs-test

**Comment by JohnTitor** [comment]
Triage: https://github.com/rust-lang/rust/blob/master/tests/ui/const-generics/const-arg-in-const-arg.rs should cover this case, closing.
